### DISCUSSION TITLE
Centralization in v2?

### DISCUSSION THREAD

**Discussion by ryanberckmans** [WalletConnect/walletconnect-monorepo#2123] [Developer Support]
Hi,

v2 looks great. However, am I misunderstanding the heightened level of centralization?

imo, what Ethereum needs is

1) to be able to use walletconnect v2 w/o centrally registering --> instead, afaict, everyone needs to register with walletconnect cloud and get an API key (projectId)

2) use coordination servers of your choice. Default to public servers or owned by walletconnect, but able to inject your own servers. This is similar to eth nodes/infura/alchemy/your own endpoints. --> instead, afaict, you need to use only WalletConnect's servers (this may be similar to v1)

3) standard connector libraries to make it easy to pull out the walletconnet QR code and link to embed it in your own UX --> instead, it seems v2 libs are designed to always pop up the web3modal (which is polished and well-made, but isn't whitelabel-able, customizable, or particularly lightweight)

Is there a way to use WalletConnect v2 without centrally registering, with my own coordination servers, and with entirely my own UX (embed QR code and link in my own app)?

**Comment by metabaseex** [comment]
I agree with you.

**Comment by ohenepee** [comment]
Let's hope they don't start adding paid API keys, access limits, etc... Cos it looks like that's where this is going.

**Comment by paintoshi** [comment]
I feared that this centralization could cause issues at some point when I started the migration to the web3modal v2. Would be very bad to block users from connecting to the wallet just because some server is down. So what I did was to add a "pre-modal" with two buttons: One using the standard injected metamask connector (no API key required) and the other one opening the wallet connect web3modal. It's one more click but feels like the safest option.

**Reply by ryanberckmans** [reply]
> I feared that this centralization could cause issues at some point when I started the migration to the web3modal v2. Would be very bad to block users from connecting to the wallet just because some server is down. So what I did was to add a "pre-modal" with two buttons: One using the standard injected metamask connector (no API key required) and the other one opening the wallet connect web3modal. It's one more click but feels like the safest option.

I'm with you, I'd hate to lock users out just because web3modal fails due to centralized servers being down or your API key being invalid. 

Another approach here may be to use an open modal like [connectkit](https://github.com/family/connectkit) or [shopify's connect-wallet](https://shopify.dev/docs/api/blockchain/components/connect-wallet.md), and then use walletconnect as a connection option inside those open modals.

**Comment by glitch-txs** [comment]
Just quoting [FAQs](https://docs.walletconnect.com/2.0/advanced/faq) here:

> We understand the desire for developers to self-host their own relay. We share this vision, and have embarked on a decentralization roadmap in order to achieve this. This summer, we will launch a permissioned network and invite a select group of partners to participate in this crucial first phase. Our objective is to make self-hosting relay a reality with the creation of the decentralized WalletConnect Network, and we appreciate your patience as we progress in this enormous mission.

**Comment by finessevanes** [comment]
Hey @ryanberckmans for your question:

>Is there a way to use WalletConnect v2 without centrally registering, with my own coordination servers, and with entirely my own UX (embed QR code and link in my own app)?

I'm going to +1 @glitch-txs [response](https://github.com/orgs/WalletConnect/discussions/2123#discussioncomment-6048353).
